feat: modularize terraform plans

jedel1043 opened the pull request:
This PR restructures the repository to offer modular components instead of a full HPC cluster.

### Changes

Instead of having a "root module", this moves everything into individual modules that can be deployed as contained units instead. The vision of this is that users pick and choose which components they need, then build their ideal deployment from those components.

Right now this only implements the "core" module, which is just slurmctld + slurmdbd + slurmrestd + slurmd, but in the future we'll want to consider implementing modules such as "observability", "identity", "login", etc.

NucciTheBoss reviewed (changes requested):
Some thoughts around the naming of things, and some ideas around the `justfile`, but I'm really liking how the modularity is shaping up here. 

I'm wondering if the `core` module should really be called `core`, or if `slurm` would be a better name for the module. Reason saying is that what we're really deploying in the core module is Slurm. My issue with calling it `core`, at least to me, is that it makes me think we're deploying a complete HPC cluster that's ready to go when really we're only deploying a single component of it.

While we can run jobs as the `ubuntu` user using a base Slurm deployment - sackd, slurmctld, slurmd, slurmdbd, slurmrestd, and mysql - you're still missing necessary components such as the identity management stack and filesystem that typically come with a complete HPC cluster.

I think that `slurm` would be a better name for the `core` module. This then allows us to tightly scope the names of the various resources for deploying the charms such as `controller`, `compute`, `rest-api`, `database`, and `backend-database` rather than use resource names such as `slurm-controller`, `slurm-compute`, `slurm-rest-api`, `slurm-database`, and `slurm-backend-database`.

Let me know if you have any comments or questions!

jedel1043 commented on justfile:
```diff
@@ -15,44 +15,35 @@
-# initialize charmed hpc deployment plans
-init:
-  #!/usr/bin/env bash
-  echo "{{bold}}Initializing plan...{{normal}}"
-  tofu init
-
 # format charmed hpc deployment plans
 fmt:
   tofu fmt -recursive
 
 # check that charmed hpc deployment plans are valid
-check: init
+check:
   #!/usr/bin/env bash
   tofu fmt -check -recursive
-  echo "{{bold}}Validating plan...{{normal}}"
-  tofu validate
+  for module in $(find {{modules}} -maxdepth 1 -mindepth 1 -type d -not -path '*/.*'); do
+    tofu -chdir=$module init -backend=false
+    tofu -chdir=$module validate
+  done
```
I can split it on two actions. However, I disagree with this:

> Also, if we roughly follow the way the justfile is implemented on the .github repository, we could make it really easy for people to use the example Terraform plans.

This works there because we're supposed to directly use the Terraform plan. However, here we want users to create their own custom plans using our modules, so IMO we should prioritize tooling for maintenance over external usage. This is important to consider because if we prioritize external usage, we would change things like `tofu -chdir=$module init -backend=false` to `tofu -chdir=$module init`, which make it easier to use externally, but it doesn't work if you don't have your Azure account set, so it's harder to use on CI.

NucciTheBoss replied:
Makes sense for me. We can keep the justfile focused on development, but then have documentation that shows folks how to use the modules to compose their own cluster deployment.

jedel1043 commented on modules/core/main.tf:
```diff
@@ -0,0 +1,128 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+data "juju_model" "this" {
```
It's just easy to remember, and since it's just an internal name I prefer to use easy to remember names

NucciTheBoss replied:
Sounds good. We can keep the name then. I kinda like `this` since it effectively treats `juju_model` as the primary resource that all the charms are attached to. I could be a Python elitist who is like, "NO WE MUST USE GLORIOUS `self`," but I think the name is fine :rofl:

jedel1043 commented:
Ok, renamed a couple of things around and added sackd to the deployment. Most of the renames are internal names, so I just changed them to the Slurm name for the nodes.

NucciTheBoss commented on examples/core/main.tf:
```diff
@@ -0,0 +1,61 @@
+  app_name        = "mysql"
+  channel         = "8.0/stable"
+  units           = 1
+}
```
> I didn't include it because you could also want to use the mysql router (even if it doesn't work right now) instead of the database directly

Works for me then. I'd imagine folks will have their own ideas about where they want MySQL to be deployed, so as long as we provide the router once `mysql-router` supports Noble, we'll be in good shape.

NucciTheBoss commented:
> Huh, I have the complete opposite perspective on this (maybe because most Rust crates follow this convention). For me, "core" means "a minimal package of software", which in my head is slurmctld + slurmdbd + slurmrestd + slurmd for Slurm.

Therein lies the issue; `core` can mean different things to different people based on their background and prior knowledge. `core` means one thing to me with my background in HPC & system administration, while it means something different to you with your low-level, Rust programming background :sweat_smile:

I agree `core` _could be_ a good name for a module that provides the bare essentials, I think a name like `slurm` or `slurm-core` makes things much less ambiguous about what your actually going to get if you use this Terraform module. `core` can include a lot of things like the filesystem and identity stack while `slurm` makes it very clear that you're going to get Slurm (at least to me)

> I named things this way so that we won't have to rename too many things if we decide to support other workload managers in the future. If we want to go all in on Slurm, I can rename things to their corresponding slurm-isms, I just need confirmation for this.

Each workload scheduler has their own peculiarities, so I'd say that we just lean into Slurm now. Slurm is by far the most popular, so we'd want strong reasoning for why we'd need to develop an entirely new set of charms lol.

NucciTheBoss commented on modules/slurm-core/main.tf:
What if we just called this module `slurm` instead of `slurm-core`? That way people don't think that we're hiding some bigger, better Slurm from them. _Maybe a non-existent scenario, but something to think about._

NucciTheBoss reviewed (changes requested):
Just one small suggestion and then I'm ready to merge :sunglasses:

NucciTheBoss reviewed (approved):
LGTM! Nice work! :shipit:
